### PR TITLE
Implementation of new scrutiny - replicas check (Part-1)

### PR DESCRIPTION
**Pull request by reddycharan**
Descriptions of the changes in this PR:

- this is implementation of scrutiny mentioned in
https://github.com/apache/bookkeeper/blob/master/site/bps/BP-34-cluster-metadata-checker.md
- In this commit, urLedgersElapsedRecoveryGracePeriod check is added to placementPolicyCheck.
- new scrutiny is added replicasCheck and it is disabled by default.
- this is first part of replicas check, in this part currently metrics are not added and
also delayed replicas check for potentially faulty ledgers (failure in trying to get
info. from Bookie of the ensembles of the ledger) is not added.
- have to add more failure test scenarios for replicas check logic.

### CODE REVIEW

**Comment by reddycharan** [comment]
rerun tests

**Comment by reddycharan** [comment]
run integration tests

**Comment by reddycharan** [comment]
run bookkeeper-server replication tests

**Comment by reddycharan** [comment]
run bookkeeper-server bookie tests

**Comment by reddycharan** [comment]
rebuild java8

**Comment by reddycharan** [comment]
run bookkeeper-server client tests

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java [hunk 1]
```diff
@@ -1072,6 +1214,366 @@ public void processResult(int rc, String s, Object obj) {
...
+
+            LedgerMetadata metadata = metadataVer.getValue();
+            if (!metadata.isClosed()) {
+                LOG.debug("Ledger: {} is not yet closed, so skipping the replicas check analysis for now",
+                        ledgerInRange);
+                mcbForThisLedgerRange.processResult(BKException.Code.OK, null, null);
+                return;
+            }
+
+            if (metadata.getLastEntryId() == -1) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java]: Nit: is there any available constant for  '-1'?

**Comment by reddycharan** [comment]
@sijie, @eolivelli and @jvrao requesting review

**Comment by eolivelli** [comment]
@reddycharan Overall looks good
just one question before giving my final +1:

do we need to add tests around WriteHandleAdv ? I mean about a ledger with temporary 
 holes.
With WriteHandleAdv  you can write entries, for instance from entryId=100 to entryId=0 and in the mean time you will have holes on the bookies.
The LAC won't be 100 until the entry with id = 0 has been acknowledged and this is expected, but how will this scrutiny behave ?

**Comment by reddycharan** [comment]
@eolivelli , As I mentioned in the BP - https://github.com/apache/bookkeeper/blob/master/site/bps/BP-34-cluster-metadata-checker.md, this checker is limited to ledgers which are write closed / fenced. Being said that I should add more negative test scenarios ("have to add more failure test scenarios for replicas check logic" mentioned in the PR description), and I will add them in the next part of this BP/change.

> High Level Logic

> Things would get complicated analyzing ledgers which are not closed because of several reasons, viz., unable to know lastEntryId by reading ZK metadata, possibility of change in ensemble because of write failure to a bookie, and other subtleties in dealing with last unclosed segment of the ledger. So for the sake of simplicity this checker should be limited to ledgers which are write closed/fenced.

**Comment by eolivelli** [comment]
run integration tests

**Comment by reddycharan** [comment]
rebased and squashed. Once tests pass, will merge this PR. 

Thanks @jvrao , @eolivelli and @sijie for the review.

**Comment by reddycharan** [comment]
rebuild java11

**Comment by reddycharan** [comment]
run bookkeeper-server replication tests

**Comment by reddycharan** [comment]
rebuild java11

**Comment by reddycharan** [comment]
rebuild java8

**Comment by reddycharan** [comment]
@sijie not sure what is the issue with these runs - Build (Java 8) and Build (Java 11). It appears build has succeeded ("BUILD SUCCESS") but finally it is failing with the following message 

`Waiting for Jenkins to finish collecting data
channel stopped
Setting status of 4cc24094f938e547b0fc2c126697b6fdabdf2068 to FAILURE with url https://builds.apache.org/job/bookkeeper_precommit_pullrequest_java8/2384/ and message: 'FAILURE
 '
Using context: Jenkins: Build (Java 8) (trigger via `rebuild java8`)
Finished: UNSTABLE`

shall i proceed with PR merge?

**Comment by eolivelli** [comment]
@reddycharan the build is usually SUCCESS but there is a test failure that is not a flake.
There is a pending patch for it.

As soon as we commit that patch, you can do it if you want (I don't have my laptop now), the master is in good shape and you will get all green lights

**Comment by eolivelli** [comment]
I am talking about https://github.com/apache/bookkeeper/pull/2125

**Comment by sijie** [comment]
@reddycharan I rebased this PR to latest master.

**Comment by reddycharan** [comment]
run bookkeeper-server tls tests

**Comment by reddycharan** [comment]
run integration tests

**Comment by reddycharan** [comment]
run bookkeeper-server replication tests

**Comment by eolivelli** [comment]
We are good to go! 
Great work Charan

**Pull request merged**